cloudinit/stages.py

```diff
@@ -395,6 +395,15 @@ def _store_rawdata(self, data, datasource):
         # Raw data is bytes, not a string
         if data is None:
             data = b''
```

Comment:
The intent of the 'rawdata' storage was to store bit-for-bit the data that was presented in the datasource. At least that was the intent for user-data, so that if cloud-init got a blob that it didn't understand, something else could read that blob from here. (/var/lib/cloud/instance/userdata) cloud-init would have then provided that data in a consistent place across different cloud providers... and the reader could then just look there.

So from that perspective, we don't want to serialize data in any way here.  We want to write the raw bytes that we found.

The problem with that is that the raw bytes have not been stored in openstack (we decoded json into ds.vendordata_raw, and then in some cases even took a subset of it)... very much not _raw.

It feels like the typeerror here is actually correct. :cry: but we have to adjust what datasources are storing in vendordata or userdata_raw.